stan-dev/stan
# Feature/updated stepsize adapt target

#### Submission Checklist


#### Summary

Addresses #2789 by updating the acceptance statistic used in the step size adaptation.

#### Intended Effect

Yields more effective samples per gradient evaluation for models with dimensions more than O(10).  Exact threshold depends on correlations in the target density.

This new branch considers each state as a proxy proposal but weights the corresponding acceptance probabilities by the state weight, e^{-H(q, p)}. In this way those states that are likely to be chosen have a greater influence on the proxy acceptance statistic, and any state with vanishing weight is not considered at all. This will always lead to a higher proxy acceptance stat for a given step size, which will cause the adaptation to push toward higher step sizes and, for a fixed integration time, cheaper numerical trajectories.

#### How to Verify

Compare effective sample size per gradient evaluation against develop.

See also the suite of test results in https://discourse.mc-stan.org/t/request-for-volunteers-to-test-adaptation-tweak/9532/28?u=betanalpha (make sure to scroll the code window -- it's long).

#### Side Effects

Pushing the step size to larger values makes the generalized No-U-Turn criterion less accurate. Combined with the multiplicative trajectory expansion this can result in a heavy tail of longer numerical trajectories, especially in lower dimensions.

#### Documentation

Inline.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Michael Betancourt

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
